File: tests/unit/wav-encoder.test.ts
import { describe, expect, it } from "bun:test";
import { WavEncoder } from "../../src/adapters/persistence/WavEncoder";

describe("WavEncoder", () => {
  it("should encode PCM16 bytes into valid RIFF WAV structure", () => {
    // 16000 Hz, 1 channel, 1000 samples = 2000 bytes = 1/16 second (62.5ms)
    const pcmData = new Uint8Array(2000);
    for (let i = 0; i < pcmData.length; i++) {
      pcmData[i] = (i * 3) % 256;
    }

    const wavBuffer = WavEncoder.encodePcm16(pcmData, 16000);

    // Total size should be 44 header bytes + 2000 data bytes = 2044 bytes
    expect(wavBuffer.length).toBe(2044);

    // Header validation
    expect(wavBuffer.subarray(0, 4).toString("ascii")).toBe("RIFF");
    expect(wavBuffer.readUInt32LE(4)).toBe(2000 + 36);
    expect(wavBuffer.subarray(8, 12).toString("ascii")).toBe("WAVE");
    expect(wavBuffer.subarray(12, 16).toString("ascii")).toBe("fmt ");
    expect(wavBuffer.readUInt32LE(16)).toBe(16); // Subchunk1Size
    expect(wavBuffer.readUInt16LE(20)).toBe(1); // AudioFormat: PCM
    expect(wavBuffer.readUInt16LE(22)).toBe(1); // NumChannels: 1
    expect(wavBuffer.readUInt32LE(24)).toBe(16000); // SampleRate
    expect(wavBuffer.readUInt32LE(28)).toBe(32000); // ByteRate = 16000 * 1 * 2
    expect(wavBuffer.readUInt16LE(32)).toBe(2); // BlockAlign
    expect(wavBuffer.readUInt16LE(34)).toBe(16); // BitsPerSample
    expect(wavBuffer.subarray(36, 40).toString("ascii")).toBe("data");
    expect(wavBuffer.readUInt32LE(40)).toBe(2000); // Subchunk2Size

    // Data payload matches
    expect(wavBuffer.subarray(44).equals(Buffer.from(pcmData))).toBe(true);
  });

  it("should accurately calculate duration in ms", () => {
    // 24000 Hz, 16-bit mono -> 48000 bytes per second
    const oneSecBytes = 48000;
    const duration = WavEncoder.calculateDurationMs(oneSecBytes, 24000);
    expect(duration).toBe(1000);

    const halfSecBytes = 24000;
    expect(WavEncoder.calculateDurationMs(halfSecBytes, 24000)).toBe(500);
  });
});

File: android/app/src/test/java/com/feynmanlive/app/audio/WavEncoderTest.kt
package com.feynmanlive.app.audio

import org.junit.Assert.assertEquals
import org.junit.Test
import java.nio.ByteBuffer
import java.nio.ByteOrder

class WavEncoderTest {

    @Test
    fun `encodes PCM16 into valid WAV structure with correct headers`() {
        val pcmData = ByteArray(3200) // 100ms at 16kHz 16-bit mono
        for (i in pcmData.indices) {
            pcmData[i] = (i % 128).toByte()
        }

        val wavBytes = WavEncoder.encodePcm16(pcmData, sampleRate = 16000, channels = 1)

        assertEquals(3200 + 44, wavBytes.size)

        // Verify RIFF and WAVE magic headers
        assertEquals('R'.code.toByte(), wavBytes[0])
        assertEquals('I'.code.toByte(), wavBytes[1])
        assertEquals('F'.code.toByte(), wavBytes[2])
        assertEquals('F'.code.toByte(), wavBytes[3])

        assertEquals('W'.code.toByte(), wavBytes[8])
        assertEquals('A'.code.toByte(), wavBytes[9])
        assertEquals('V'.code.toByte(), wavBytes[10])
        assertEquals('E'.code.toByte(), wavBytes[11])

        val buffer = ByteBuffer.wrap(wavBytes).order(ByteOrder.LITTLE_ENDIAN)
        val chunkLen = buffer.getInt(4)
        assertEquals(3200 + 36, chunkLen)

        val audioFormat = buffer.getShort(20)
        assertEquals(1.toShort(), audioFormat) // PCM

        val numChannels = buffer.getShort(22)
        assertEquals(1.toShort(), numChannels)

        val sampleRate = buffer.getInt(24)
        assertEquals(16000, sampleRate)

        val bitsPerSample = buffer.getShort(34)
        assertEquals(16.toShort(), bitsPerSample)

        val dataLen = buffer.getInt(40)
        assertEquals(3200, dataLen)
    }

    @Test
    fun `calculates duration in ms accurately`() {
        // 16000 samples/sec * 2 bytes/sample = 32000 bytes/sec
        // 32000 bytes = 1000 ms
        val duration1s = WavEncoder.calculateDurationMs(32000L, 16000, 1)
        assertEquals(1000L, duration1s)

        val durationHalfSec = WavEncoder.calculateDurationMs(16000L, 16000, 1)
        assertEquals(500L, durationHalfSec)
    }
}
